### ISSUE TITLE
Can't `hashComponents` if bindings have free variables

### ISSUE THREAD

**Issue by pchiusano** [unisonweb/unison#4748]
```
SearchIndex.FullText.util.hashes : Nat -> [Nat] -> [Text] -> [SuffixHash Text]
SearchIndex.FullText.util.hashes hashingThreshold windowSizes words = 
  goWord : Text -> {Stream (SuffixHash Text)} ()
  goWord word = 
    todo "tricky" 

  go : [Text] -> [SuffixHash Text]
  go words = Stream.toList! (
    (do List.foreach words goWord) : '{Stream (SuffixHash Text)} ()
  ) 
  go words
```

```
  I found and typechecked these definitions in ~/unison/scratch.u. If you do an `add` or `update`,
  here's how your codebase would change:
  
    ⍟ These new definitions are ok to `add`:
    
      SearchIndex.FullText.util.hashes : Nat -> [Nat] -> [Text] -> [SuffixHash Text]

Encountered exception:
can't hashComponents if bindings have free variables:
  ["User \"words\""]
  ["User \"SearchIndex.FullText.util.hashes\"","User \"go\"","User \"goWord\"","Float","Float-1"]
CallStack (from HasCallStack):
  error, called at src/Unison/Hashing/V2/ABT.hs:77:11 in unison-hashing-v2-0.0.0-G2HCpgVgyjV7tBghoMJbOY:Unison.Hashing.V2.ABT
```

**Comment by benwaffle** [comment]
I'm seeing this as well, when trying to `run` a function. Seems like everything typechecks properly.

**Comment by benwaffle** [comment]
This happened because of a missing variable. The error message is very confusing.

**Comment by dfreeman** [comment]
I just hit this as well, in what looks like a very similar case to what @pchiusano posted above where I needed to ascribe a specific type to a `do` expression that closed over an in-scope variable. This watch expression reliably reproduces the issue for me:

```
> let
  x = 123
  ((do x) : 'Nat)
```

While this, without the type ascription, seems fine:

```
> let
  x = 123
  do x
```

**Comment by lJoublanc** [comment]
I've closed another issue which I think is the same as this.

I used an llm to generate the analysis below:


**Describe and demonstrate the bug**

When loading a `.u` scratch file that contains a `test>` watch expression (or terms evaluated within doc blocks), UCM crashes with an uncaught exception in `Unison.Hashing.V2.ABT.hashComponents`:

```text
Encountered exception:
can't hashComponents if bindings have free variables:
  ["User \"preds\""]
  ["User \"Adjoint.undual\"","User \"Dimensionless.coerce\"","User \"Dimensionless.coerce.doc\"","User \"Dimensionless.doc\"","User \"Dimensionless.map\"","User \"Dimensionless.map.doc\"","User \"Dimensionless.map2\"","User \"Dimensionless.map2.doc\"","User \"Dual.doc\"","User \"README\"","User \"dimensionless\"","User \"examples.rollercoaster\"","User \"withAdjoint\"","User \"withAdjoint.doc\"","User \"withAdjoint.tests.hadamardFiniteDifference\""]
CallStack (from HasCallStack):
  error, called at src/Unison/Hashing/V2/ABT.hs:130:11 in unison-hashing-v2-0.0.0-2jgI31DfNzBLXeVQYi9pgR:Unison.Hashing.V2.ABT
HasCallStack backtrace:
  collectBacktraces, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:169:13 in ghc-internal:GHC.Internal.Exception
  toExceptionWithBacktrace, called at libraries/ghc-internal/src/GHC/Internal/IO.hs:260:11 in ghc-internal:GHC.Internal.IO
  throwIO, called at src/Ki/Internal/Scope.hs:198:25 in ki-1.0.1.2-G0k6FwDdV3TDP6J4dPCKAy:Ki.Internal.Scope
```

The file successfully passes typechecking (`unison_typecheck` / type inference), but crashes during UCM's watch expression evaluation / component hashing phase.

### Transcript Reproduction

``` ucm :hide
scratch/main> builtins.merge
```

Here we define an ability, a handler, and a function where a variable name (`preds`) is used both as a parameter in a local helper (`mse preds targets = ...`) and as a local `let` binding inside a delayed `do` block passed to the handler (`result = withAdjoint do ... preds = ...`).

``` unison
ability Adjoint expr where
  adjoint : Dual expr -> Dual expr ->{Adjoint expr} Dual expr

type Dual expr = Dual Nat expr

withAdjoint : '{Adjoint matrix} r -> r
withAdjoint prog =
  go = cases
    { r } -> r
    { Adjoint.adjoint a b -> cont } ->
      handle cont a with go
  handle prog() with go

buggy : Nat -> (Nat, Nat)
buggy n =
  -- 1. Helper function where `preds` is a parameter
  mse preds targets =
    preds + targets

  -- 2. Local step function with a nested `do` block passed to an effect handler
  step w =
    res = withAdjoint do
      -- 3. `preds` is bound as a local variable inside the delayed `do` block
      preds = w + 1
      loss = mse preds 3
      Adjoint.adjoint w loss
    -- 4. Tuple destructuring pattern match
    (loss, w') = res
    (loss, w')

  step n
```

When evaluated via a watch expression (`>` or `test>`), UCM must hash the components via `hashComponents` for the sandbox runtime. If the ABT desugaring pass fails to bind the inner identifier, `hashComponents` throws:

``` ucm :error
scratch/main> load scratch.u
```

**Screenshots**
N/A

**Environment:**
 - `ucm --version`: `unison version: release/1.4.0 (built on 2026-08-19)`
 - OS/Architecture: `Linux (Ubuntu 24.04 LTS, x86_64, kernel 6.8.0-138-generic)`

**Additional context**

### Root Cause Analysis
1. **Scoping & ABT Construction**: The variable name `preds` is used in two places in the same enclosing definition:
   - As a parameter in a local helper function `mse preds targets = ...`.
   - As a `let` binding inside an inner `do` block passed to a higher-order handler function `withAdjoint do ...`.
   During lambda-lifting / AST desugaring of the nested `do` block and pattern destructuring (`(loss, w') = res`), the identifier `User "preds"` is not correctly resolved to its inner scope in the Abstract Binding Tree (ABT).
2. **Delayed Failure**: The pure type checker succeeds because types match locally. However, when a `test>` or `>` watch expression is present, UCM attempts to compile and run the scratch components in the runtime sandbox. This requires computing cryptographic content hashes via `Unison.Hashing.V2.ABT.hashComponents`, which asserts that all terms have zero unbound free variables and crashes when `User "preds"` is encountered.
3. **Workarounds**:
   - Renaming the variable in either the parameter list or the inner `let` binding eliminates the conflict and allows compilation/execution.
   - Removing or commenting out `test>` watch expressions bypasses the sandbox execution pass.
